### PR TITLE
Implement fast summarization

### PR DESCRIPTION
**Pull request by anvacaru**
fixes: #266

**Changes:**
- add a new `--init-node-from` flag that takes a JSON file containing the summary of a deployment process. Takes the accounts obtained after going through the summary and inserts them in the `<accounts>` cell of the initial node of the KCFG.

- Extract the logic that reads the JSON summary files in `foundry_summary()` to its own `read_summary()` function.

- Refactor `DeploymentSummary` class:
     - extract the logic that processes a summary entry into its own `SummaryEntry` class.
     -  refactoring the `DeploymentSummary.add_cheatcode(dict)` that would process a summary entry and create the Solidity code for the cheat codes required to update the state to `DeploymentSummary.extend(SummaryEntry)`.
 
 - Add `summary_to_account_cells(dict)` that goes through the summary JSON object and creates a list of `<account>` cells. 
 
**Minor changes:**
- replace occurrences of  `KApply(.Map)` to `map_empty()`
- remove unused `_write_cfg` function

Next steps:
 1. Apply the state diff changes over an account that has symbolic fields  like `{'balance': KVariable('ACCT_X_BAL'), 'nonce': KVariable('ACCT_X_NONCE') , ...}` instead of `{'balance': 0, 'nonce': 0, 'code': '', 'storage': {}}`
 2. Insert the set of account addresses in `<accessedAccounts>`,`<touchedAccounts>` (not sure actually?)
 3. Currently, the storage updates from the summary are placed only in `<storage>`, apply them in `<origStorage>` as well.

### CODE REVIEW

File: src/kontrol/options.py
Comment (tothtamas28) [inline src/kontrol/options.py:31]: Will this eventually be transformed into a `Deploymentsummary`? If so, consider doing it right in `__init__` (or even earlier), as the sooner you get rid of data that's not validated / whose type is not statically known the better,

File: src/kontrol/prove.py
Comment (tothtamas28) [inline src/kontrol/prove.py:476]: This seems like the central feature in this PR, please add a (parametric) unit test that exercises it. This would catch changes to the `summary` schema.

File: src/kontrol/prove.py
Comment (tothtamas28) [inline src/kontrol/prove.py:518]: Is extending the list safe, i.e. is it possible for summary accounts to somehow interfere with the other two accounts (e.g. same address)? If so, consider doing some validation.

File: src/kontrol/prove.py
Comment (tothtamas28) [reply src/kontrol/prove.py:476]: It would be good to convert `summary` into a statically typed data structure soon (or right) after reading the data from disk.

File: src/kontrol/options.py
Comment (anvacaru) [reply src/kontrol/options.py:31]: No, the `DeploymentSummary` is used only in the functionality of the `foundry_summary` function to generate the Solidity contract based on the `summary` dict.
But I get the idea and I can process the `summary` into a list of `SummaryEntry`s.

File: src/kontrol/deployment.py
Comment (anvacaru) [reply src/kontrol/deployment.py]: Implemented in https://github.com/runtimeverification/kontrol/pull/271/commits/751d5d70895e7db0f3ee38663ff8dc25a8819858.

File: src/kontrol/foundry.py
Comment (anvacaru) [reply src/kontrol/foundry.py]: Done in https://github.com/runtimeverification/kontrol/pull/271/commits/2ba5fe5f39b7313e9b058f4183c37ecbbb35ded3.

File: src/kontrol/options.py
Comment (anvacaru) [reply src/kontrol/options.py]: Now, `summary` is `list[SummaryEntry] | None`.

File: src/kontrol/prove.py
Comment (anvacaru) [reply src/kontrol/prove.py]: Added a unit test in https://github.com/runtimeverification/kontrol/pull/271/commits/55155001459a65c4c4db8ef038cb7c8f76ddc8c6, and`summary` is now a `list[SummaryEntry]`.

File: src/kontrol/deployment.py
Comment (anvacaru) [reply src/kontrol/deployment.py:54]: Yes, this handles the generation of Solidity code. There currently is an [integration test in place].(https://github.com/runtimeverification/kontrol/blob/master/src/tests/integration/test_foundry_prove.py#L507)

File: src/kontrol/deployment.py
Comment (anvacaru) [reply src/kontrol/deployment.py]: Done in https://github.com/runtimeverification/kontrol/pull/271/commits/76e84517fbdd026166eedb63f972b932cf1c7cc7.

File: src/kontrol/prove.py
Comment (anvacaru) [reply src/kontrol/prove.py]: Moved it in https://github.com/runtimeverification/kontrol/pull/271/commits/76e84517fbdd026166eedb63f972b932cf1c7cc7.

**Comment by ehildenb** [comment]
It would be good to test this on the Optimism codebase where the summaries are being used. @JuanCoRo can you confirm that this works for your needs?

**Comment by palinatolmach** [comment]
Looks good to me too! I think @anvacaru tested it on the engagement proof of interest, but I'll wait for an additional confirmation before approving and merging.

**Review by palinatolmach — APPROVED** [review]

**Pull request merged**
